test/jdk/jdk/crac/DaemonAfterRestore.java

```diff
@@ -0,0 +1,107 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import jdk.crac.*;
+import jdk.test.lib.crac.CracBuilder;
+import jdk.test.lib.crac.CracProcess;
+import jdk.test.lib.crac.CracTest;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @test
+ * @summary All afterRestore's should complete, even if there are only daemon threads (in case one of the afterRestore's finally creates non-daemon thread that will be responsible to keep VM alive)
+ * @library /test/lib
+ * @build DaemonAfterRestore
+ * @run driver jdk.test.lib.crac.CracTest
+ */
+public class DaemonAfterRestore implements CracTest {
+    static final String MAIN_THREAD_FINISH = "main thread finish";
+    static final String AFTER_RESTORE_MESSAGE = "after restore finish";
+
+    @Override
+    public void test() throws Exception {
+        CracBuilder builder = new CracBuilder().captureOutput(true);
+
+        CompletableFuture<?> firstOutputFuture = new CompletableFuture<Void>();
+        CracProcess checkpointProcess = builder.startCheckpoint().watch(
+            outline -> {
+                System.out.println(outline);
+                if (outline.equals(MAIN_THREAD_FINISH)) {
+                    firstOutputFuture.complete(null);
+                }
+            },
+            errline -> {
+                System.err.println("ERROR: " + errline);
+                firstOutputFuture.cancel(false);
+            });
+        firstOutputFuture.get(10, TimeUnit.SECONDS);
+        builder.checkpointViaJcmd();
+        checkpointProcess.waitForCheckpointed();
+
+        builder.startRestore().waitForSuccess()
+            .outputAnalyzer().shouldContain(AFTER_RESTORE_MESSAGE);
+    }
+
+    @Override
+    public void exec() throws RestoreException, CheckpointException {
+        CountDownLatch start = new CountDownLatch(1);
+        CountDownLatch finish = new CountDownLatch(1);
+        Thread workerThread = new Thread(() -> {
+            System.out.println("worker thread start");
+            start.countDown();
+            try {
+                finish.await();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            System.out.println("worker thread finish");
+        });
+        workerThread.setDaemon(false);
+        workerThread.start();
+
+        try {
+            start.await();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+
+        Resource resource = new Resource() {
+            @Override
+            public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+                finish.countDown();
+            }
+            @Override
+            public void afterRestore(Context<? extends Resource> context) throws Exception {
+                Thread.sleep(3000);
```

Comment:
Could we avoid extending the testsuite run by 3 seconds? I know we're trying to assert that 'nothing happens' rather then replacing waiting for an event by a timed wait. If we can't check that the destroy VM thread had a chance to work, I suggest at least reducing this (say 100ms?).

Reply:
100ms does not look as a good candidate (comparable to a single scheduling period). So 3 sec does not look to bad to avoid interrmittent false positive because sleep() finished before VM has terminated.